### ISSUE TITLE
BlocBuilder is changing his state but GridView still displays old data

### ISSUE THREAD

**Issue by IsidorosT** [felangel/bloc#3049]
I have a State Object for my Settings Bloc: 
```
class SettingsState{
  late bool UnsavedSettings;
  late bool UnsavedUserGroups;
  late String SelectedGroupName;
  late List<GroupSetting> ActiveSettings;
  SettingsState(bool unsavedSettings,bool unsavedUsergroups,String selectedGroupName, List<GroupSetting> activeSettings){
    UnsavedSettings = unsavedSettings;
    UnsavedUserGroups = unsavedUsergroups;
    SelectedGroupName = selectedGroupName;
    ActiveSettings = new List.from(activeSettings);
  }
}
```
And i have a widget that gets the state from bloc builder and displays it into a GridView builder:
```
Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    return BlocBuilder<SettingsBloc, SettingsState>(builder: (_, settingsState) { 
              ...
              Expanded(
                   child:GridView.builder(
                      shrinkWrap: true,
                      primary: false,
                      addAutomaticKeepAlives: false,
                      padding: const EdgeInsets.all(20),
                      //crossAxisCount: 2,

                      gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                        crossAxisCount: 2,
                        childAspectRatio: 4.5,
                        crossAxisSpacing: 15,
                        mainAxisSpacing: 10,
                        mainAxisExtent: 135,
                      ),
                      itemCount: settingsState.ActiveSettings.length,
                      itemBuilder: (ctx, index) {
                        return Visibility(
                          visible: settingsState.ActiveSettings[index].Visible,
                            child: SettingCard(settingsState.ActiveSettings[index], SettLogic));
                      }

                  )
             )
              ...
     }
}
```

List did not refresh but if i scroll down enough to hide it or change my app's tab the new information is displayed as should be.
I have also another Bloc wich handling a ListView with very similar way and i dont have that problem.

**Comment by IsidorosT** [comment]
Hi @felangel and thank you for your response.
Here is a little project that i think reproduces the error above
[repro1.zip](https://github.com/felangel/bloc/files/7672074/repro1.zip)
I have a bloc that updates the list of grid view but it doesnt refresh visually

**Comment by IsidorosT** [comment]
Any Thoughts?

**Comment by narcodico** [comment]
Hi @IsidorosT 👋

Could you provide a github repository with a more simplified example of your issue? Your current sample needs A LOT of refactoring to get it in a good spot. Please keep it simple just to exemplify the issue.

I'd also advise you to read more about the basic of dart/flutter since you seem to have a C# style approach to your code, especially when it comes to object/prop initializations. Please also read more about the basics of bloc since it seems you're missing on a lot of the fundamentals; you can start [here](https://bloclibrary.dev/#/gettingstarted)

**Comment by felangel** [comment]
Closing for now but feel free to comment with a link to a GitHub repo or DartPad that contains a minimal reproduction sample and I'm happy to take a closer look, thanks!

**Comment by IsidorosT** [comment]
I tried to simplify the reproduction sample. 
As for the documentation that you suggested i have used it a lot.
Maybe i miss something basic but i have read that blocbuilder redraw widget when state changed in builder function. 

So i simplified it [here](https://github.com/IsidorosT/ReproductionSampleForBlocIssue)

The sample has one TestBloc that has a list.
In main widget we have a gridview that prints state's list in a widget and two buttons that change the state causing other values to be inserted in new state's list

I also inserted a print in event state that it shows that the new state is emited but grid view does not refresh.
